### PR TITLE
Add multinomial distribution to sympy.stats

### PR DESCRIPTION
**Pull request by dr-leo**
No unit test yet, but doctest.

### CODE REVIEW

**Comment by asmeurer** [comment]
You need to add the class to test_args (see the Travis tests).

**Comment by asmeurer** [comment]
Please add some tests for this.

File: sympy/stats/frv_types.py [hunk 1]
```diff
@@ -251,6 +259,60 @@ def Binomial(name, n, p, succ=1, fail=0):
...
+        n = as_int(n)
+        coeff = multinomial_coefficients(len(p), n)
+        return {k: v * prod(map(pow, p, k)) for k, v in coeff.items()}
+
+
+def Multinomial(name, n, p):
+    """
+    Create a Finite Random Variable representing a multinomial distribution.
+    p: list-like vector of probabilities. p[i] must add up to 1.
+
```
Comment (asmeurer) [inline sympy/stats/frv_types.py]: Can you add a mathematical definition (using `.. math::`)

**Comment by Upabjojr** [comment]
This would be the first distribution acting on multi-dimensional space to be included into SymPy's master branch.

@mrocklin had some examples:
* https://sympystats.wordpress.com/2011/07/19/multivariate-normal-random-variables/
* https://github.com/mrocklin/sympy/tree/mvn_rv/

His branch has not been merged so far.

**Comment by Upabjojr** [comment]
Maybe it is advisable to test computations of:

* probability
* expectation
* variance and covariance

Also, see if the same expressions are supported with a conditional expression. I've got a feeling that this may not work.

**Comment by Upabjojr** [comment]
https://github.com/sympy/sympy/issues/11499

**Comment by dr-leo** [comment]
Thanks to all for the helpful Feedback. The branch is not yet ripe for
merging. I realize it would require some more General infrastructure
for multivariate discrete random variables.

The proposal at
https://sympystats.wordpress.com/2011/07/19/multivariate-normal-random-variables/

relates to a continuous random variable. In 2011 it was shelved due to
the then lacking Support for matrices in sympy. It does not seem to me
that this assumption still holds true today. So it might be worthwhile
revisiting that proposal.

Subject to very tight time constraints I will have a fresh look at
what would be needed to properly implement multivariate discrete RV's
in sympy.

No Need to say that any help would be appreciated.

Leo

**Comment by Upabjojr** [comment]
@dr-leo just test expressions like 
* _expectation(X)_
* _probability(X = (1,2,3))_
* _probability(X > (1,2,3))_
* _variance(X)_ 

with your distribution.

**Comment by dr-leo** [comment]
Thanks. Your examples on E, Var and = granted and should be easy to
implement once I've got a grasp of, and we agree on, an appropriate
Extension of the class hierarchy.

But what does P(x > (1,2,3) mean? N^n is not an ordered set. Do you
imply a Euclidian metric? - I am not into the theory. But I suspect
there is a Need for clarification / generalization to different
metrics.

**Comment by Upabjojr** [comment]
> Thanks. Your examples on E, Var and = granted and should be easy to implement once I've got a grasp of, and we agree on, an appropriate Extension of the class hierarchy.

I believe it can be solved by adding a for loop to those functions, i.e. nested summations (or integrals for continuous RVs). Not sure on how to store the information on the multivariate RV.

> But what does P(x > (1,2,3) mean? N^n is not an ordered set. Do you imply a Euclidian metric? 

Maybe better leave this example out.

**Comment by asmeurer** [comment]
So should X be a matrix expression, so that things like `X[0]` work (and hence `And(X[0] > 1, X[1] > 2, X[2] > 3)`)?

**Comment by dr-leo** [comment]
Yes, from a practical viewpoint element-wise comparisons make sense
beyond equality. But is this in sync with matrix comparisons in general?
- I am afraid I lack the time to review matrix theory to make an
informed judgement here.

On a separate note: The class hierarchy within stats is not that obvious
when looking at it late at night ;-)

So the more fundamental question seems to me:

1. Do we need an abstract class like MultiPSpace etc.? Or does
Product... class do the trick?

This would take me much more time than expected. The formulae for the
multinomial distribution for E, Var etc. are easy. But generalizing to
multivariate domains etc. takes more time. This is much mor difficult
than adding comparisons for matrices, tuples etc. in a given place.

I even have to look at the __new__ method first, I've never used this
before.
 
 
 
 


Am 18.08.2016 um 06:51 schrieb Aaron Meurer:

**Comment by asmeurer** [comment]
These additions can be added later. They don't need to be in the first iteration.

**Comment by Upabjojr** [comment]
> 1. Do we need an abstract class like MultiPSpace etc.? Or does Product... class do the trick?

Maybe, but better concentrate on it in the future. The class hierarchy in this module is very complicated, better have a full grasp on it before trying to edit it.

> This would take me much more time than expected. The formulae for the multinomial distribution for E, Var etc. are easy. But generalizing to multivariate domains etc. takes more time. This is much mor difficult than adding comparisons for matrices, tuples etc. in a given place.

Better concentrating on simple stuff. Multivariate domains could take years to implement.

> I even have to look at the __new__ method first, I've never used this before.

Do we have a guide for new contributors on how to write class constructors?

Anyway, expressions like `E(X+Y)` already work, so I suppose `E`, `Var`, `P` already support multivariate expression of RVs (expressions of monovariate RVs). The question is simply how to make them aware that even a single RV might be multivariate. I would try to do it as similar as possible to the way used in @mrocklin 's implementation of matrix RV.

**Comment by Upabjojr** [comment]
Anyways, if you add some tests to this PR, I think we can merge it.

**Comment by sidhantnagpal** [comment]
Can this and #11499 be closed?
#16808 seems to have resolved both. There is also related work in #16834.

**Comment by dr-leo** [comment]
Yes.

**Pull request closed without merging**